File: appkit/layout_anchor.m
#import <Appkit/NSLayoutAnchor.h>
#import <Appkit/NSLayoutConstraint.h>
#import "layout_anchor.h"

Array LayoutAnchor_ConstraintsAffectingLayout(void* ptr) {
	NSLayoutAnchor* layoutAnchor = (NSLayoutAnchor*)ptr;
	NSArray* array = [layoutAnchor constraintsAffectingLayout];
	int count = [array count];
	void** data = malloc(count * sizeof(void*));
	for (int i = 0; i < count; i++) {
		 data[i] = [array objectAtIndex:i];
	}
	Array result;
	result.data = data;
	result.len = count;
	return result;
}

bool LayoutAnchor_HasAmbiguousLayout(void* ptr) {
	NSLayoutAnchor* layoutAnchor = (NSLayoutAnchor*)ptr;
	return [layoutAnchor hasAmbiguousLayout];
}

const char* LayoutAnchor_Name(void* ptr) {
	NSLayoutAnchor* layoutAnchor = (NSLayoutAnchor*)ptr;
	return [[layoutAnchor name] UTF8String];
}

void* LayoutAnchor_Item(void* ptr) {
	NSLayoutAnchor* layoutAnchor = (NSLayoutAnchor*)ptr;
	return [layoutAnchor item];
}

void* LayoutAnchor_ConstraintEqualToAnchor(void* ptr, void* anchor) {
	NSLayoutAnchor* layoutAnchor = (NSLayoutAnchor*)ptr;
	return [layoutAnchor constraintEqualToAnchor:(NSLayoutAnchor*)anchor];
}

void* LayoutAnchor_ConstraintEqualToAnchor2(void* ptr, void* anchor, double constant) {
	NSLayoutAnchor* layoutAnchor = (NSLayoutAnchor*)ptr;
	return [layoutAnchor constraintEqualToAnchor:(NSLayoutAnchor*)anchor constant:constant];
}

void* LayoutAnchor_ConstraintGreaterThanOrEqualToAnchor(void* ptr, void* anchor) {
	NSLayoutAnchor* layoutAnchor = (NSLayoutAnchor*)ptr;
	return [layoutAnchor constraintGreaterThanOrEqualToAnchor:(NSLayoutAnchor*)anchor];
}

void* LayoutAnchor_ConstraintGreaterThanOrEqualToAnchor2(void* ptr, void* anchor, double constant) {
	NSLayoutAnchor* layoutAnchor = (NSLayoutAnchor*)ptr;
	return [layoutAnchor constraintGreaterThanOrEqualToAnchor:(NSLayoutAnchor*)anchor constant:constant];
}

void* LayoutAnchor_ConstraintLessThanOrEqualToAnchor(void* ptr, void* anchor) {
	NSLayoutAnchor* layoutAnchor = (NSLayoutAnchor*)ptr;
	return [layoutAnchor constraintLessThanOrEqualToAnchor:(NSLayoutAnchor*)anchor];
}

void* LayoutAnchor_ConstraintLessThanOrEqualToAnchor2(void* ptr, void* anchor, double constant) {
	NSLayoutAnchor* layoutAnchor = (NSLayoutAnchor*)ptr;
	return [layoutAnchor constraintLessThanOrEqualToAnchor:(NSLayoutAnchor*)anchor constant:constant];
}

File: scripts/layout_anchor.py
#!env python3

from generate import Component, Property, Method, Return, Param, init_method


if __name__ == "__main__":
    w = Component(
        Type="appkit.LayoutAnchor",
        super_type='foundation.Object',
        objc_imports=['Appkit/NSLayoutAnchor', 'Appkit/NSLayoutConstraint'],
        description="a factory class for creating layout constraint objects using a fluent API",
        properties=[
            Property(name='constraintsAffectingLayout', Type='appkit.LayoutConstraint', array=True, readonly=True,
                     description='the constraints that impact the layout of the anchor'),
            Property(name='hasAmbiguousLayout', Type='bool', readonly=True, getter_prefix_is=False,
                     description='whether the constraints impacting the anchor specify its location ambiguously'),
            Property(name='name', Type='string', readonly=True,
                     description='the name assigned to the anchor for debugging purposes'),
            Property(name='item', Type='foundation.Object', readonly=True,
                     description='the layout item used to calculate the anchor\'s position'),
                     

        ],
        methods=[
            Method(
                name='constraintEqualToAnchor',
                params=[
                    Param(name='anchor', Type='appkit.LayoutAnchor'),
                ],
                return_value=Return(Type='LayoutConstraint'),
                description='returns a constraint that defines one item’s attribute as equal to another',
            ),
            Method(
                name='constraintEqualToAnchor',
                go_func_name='ConstraintEqualToAnchor2',
                params=[
                    Param(name='anchor', Type='appkit.LayoutAnchor'),
                    Param(name='constant', Type='float64', objc_param_name='constant'),
                ],
                return_value=Return(Type='LayoutConstraint'),
                description='returns a constraint that defines one item’s attribute as equal to another item’s attribute plus a constant offset',
            ),
            Method(
                name='constraintGreaterThanOrEqualToAnchor',
                params=[
                    Param(name='anchor', Type='appkit.LayoutAnchor'),
                ],
                return_value=Return(Type='LayoutConstraint'),
                description='returns a constraint that defines one item’s attribute as greater than or equal to another',
            ),
            Method(
                name='constraintGreaterThanOrEqualToAnchor',
                go_func_name='ConstraintGreaterThanOrEqualToAnchor2',
                params=[
                    Param(name='anchor', Type='appkit.LayoutAnchor'),
                    Param(name='constant', Type='float64', objc_param_name='constant'),
                ],
                return_value=Return(Type='LayoutConstraint'),
                description='returns a constraint that defines one item’s attribute as greater than or equal to another item’s attribute plus a constant offset',
            ),
            Method(
                name='constraintLessThanOrEqualToAnchor',
                params=[
                    Param(name='anchor', Type='appkit.LayoutAnchor'),
                ],
                return_value=Return(Type='LayoutConstraint'),
                description='returns a constraint that defines one item’s attribute as less than or equal to another',
            ),
            Method(
                name='constraintLessThanOrEqualToAnchor',
                go_func_name='ConstraintLessThanOrEqualToAnchor2',
                params=[
                    Param(name='anchor', Type='appkit.LayoutAnchor'),
                    Param(name='constant', Type='float64', objc_param_name='constant'),
                ],
                return_value=Return(Type='LayoutConstraint'),
                description='returns a constraint that defines one item’s attribute as less than or equal to another item’s attribute plus a constant offset',
            ),
        ]
    )
    w.generate_code()
